cluster_management/src/main/java/com/pinterest/rocksplicator/Utils.java

```diff
@@ -472,10 +477,55 @@ public static void checkStateTransitions(String fromState, String toState, Messa
     LOG.error("From " + fromState + " to " + toState + " for " + partitionName);
   }
 
+  public static String getMetaParentClusterLocation(String cluster) {
+    return "/metadata/" + cluster;
+  }
+
+  public static String getMetaParentResourceLocation(String cluster, String resourceName) {
+    return getMetaParentClusterLocation(cluster) + "/" + resourceName;
+  }
+
   public static String getMetaLocation(String cluster, String resourceName) {
-    return "/metadata/" + cluster + "/" + resourceName + "/resource_meta";
+    return  getMetaParentResourceLocation(cluster, resourceName) + "/resource_meta";
   }
 
+  public static String getMetaData(
+      final CuratorFramework zkClient,
+      final String cluster,
+      final String resourceName,
+      final int max_retries) throws Exception {
+    final String clusterMetaPath = Utils.getMetaParentClusterLocation(cluster);
+    final String resourceMetaPath = Utils.getMetaParentResourceLocation(cluster, resourceName);
+    final String metadataPath = Utils.getMetaLocation(cluster, resourceName);
+
+    final AtomicInteger retryCount = new AtomicInteger(0);
+    return RetryLoop.callWithRetry(zkClient.getZookeeperClient(), new Callable<String>() {
+      @Override
+      public String call() throws Exception {
+        retryCount.incrementAndGet();
+        try {
+          zkClient.sync().forPath(clusterMetaPath);
+          zkClient.sync().forPath(resourceMetaPath);
```

Comment:
I have two Qs: 1) can u share what does sync() do? why do we need it?
2) If we only access the overall path, why do we need to sync the parent path?

Reply:
The CuratorFramework doesn't provide a documentation on it, but it believe it performs a synchronous sync from leader to the follower this client is connected to, hence it forces the latest view on given zknode.

Reason for recursively syncing path from top to bottom, is if the zk follower is lagging behind in syncing the parent path, it will not see a newly created children node. This is what is causing the regular exceptions when resources are created a fresh per version (internal concept)